File: cpp9/sandbox/mergesort_using_deque/sort.py
import sys, typing
from collections import deque
from random import randint

def merge_sort(D):

    # basecase
    if len(D) < 2:
        return

    # split
    size = len(D)
    mid = size // 2
    L = deque()
    R = deque()
    for i in range(mid):
        L.append(D.pop())
    for i in range(size - mid):
        R.append(D.pop())

    # recurse
    merge_sort(L)
    merge_sort(R)

    # compare & insert
    while L and R:
        if L[0] > R[0]:
            D.append(R.popleft())
        else:
            D.append(L.popleft())

    # get the rest
    while L:
        D.append(L.popleft())
    while R:
        D.append(R.popleft())


if __name__ == '__main__':
    def debugger(a: list) -> None:
        printlist(a)
        d = deque(a)
        merge_sort(d)
        printlist(list(d), '(sorted)\n')

    def printlist(a: list, end: str = '') -> None:
        if len(a) > 12:
            print(a[:6], '[...]', len(a), end)
        else:
            print(a, len(a), end)

    a = []

    # pmergeme test
    if len(sys.argv) == 1:
        a = [3, 5, 9, 7, 4]
        debugger(a)
        a = [2, 11, 1, 42, 21, 1000, 6, 3, -42]
        debugger(a)
        a = [2, 11, 1, 42, 21, 1000, 6, 3, -42, 21]
        debugger(a)
    # random test
    else:
        print(sys.argv[1])
        if sys.argv[1] == 'test':
            a = [ randint(1, 100000) for _ in range(0, 3000) ]
        else:
            a = [ int(_) for _ in sys.argv[1:] ]
    debugger(a)

File: cpp9/sandbox/mergesort_using_array/sort.ts
const process = require('process')

if (process.argv.length == 2) {
	let arr = new Array()

	arr = [3, 5, 9, 7, 4]
	test(arr)
	arr = [2, 11, 1, 42, 21, 1000, 6, 3, -42]
	test(arr)
	arr = [2, 11, 1, 42, 21, 1000, 6, 3, -42, 21]
	test(arr)
}

function merge_sort(arr) {
	if (arr.length < 2)
		return

	// split
	let len = arr.length
	let mid = Math.floor(len / 2)
	let L = [] //= arr.slice(mid, len).reverse()
	let R = [] //arr.slice(0, mid).reverse()

	for (let i = 0; i < mid; i++) {
		R.push(arr.pop())
	}
	for (let i = 0; i < len - mid; i++) {
		L.push(arr.pop())
	}

	// console.log(L)
	// console.log(R)

	// recurse
	merge_sort(L)
	merge_sort(R)

	// insert
	while (L.length != 0 && R.length != 0) {

		if (L[0] < R[0]) {
			arr.push(L.shift())
		} else {
			arr.push(R.shift())
		}
	}

	// the rest
	while (L.length != 0) {
		arr.push(L.shift())
	}
	while (R.length != 0) {
		arr.push(R.shift())
	}
}

function print_array(arr) {
	process.stdout.write('[');
	for (let i = 0; i < arr.length; i++) {
		let n = arr[i]
		let len = arr.length
		let sep = i == len ? ' ' : ', '
		process.stdout.write(arr[i] + sep);
	}
	process.stdout.write('] ');
}

function print_result(arr, end='') {
	let len = arr.length
	if (len > 12) {
		print_array(arr.slice(0, 6))
		console.log('[...]', len, end)
	} else {
		print_array(arr)
		console.log(len, end)
	}
}

function test(arr) {
	print_result(arr)
	merge_sort(arr)
	print_result(arr, '(sorted) \n')
}
